horreum-backend/src/test/java/io/hyperfoil/tools/horreum/svc/SchemaServiceTest.java

```diff
@@ -33,21 +30,176 @@
-public class SchemaServiceTest extends BaseServiceTest {
+class SchemaServiceTest extends BaseServiceTest {
 
    @Inject
    ServiceMediator serviceMediator;
 
    @org.junit.jupiter.api.Test
-   public void testValidateRun() throws IOException, InterruptedException {
+   void testGetSchema() {
+      String schemaUri = "urn:dummy:schema";
+      Schema schema = createSchema("Dummy schema", schemaUri);
+      assertNotNull(schema);
+      assertTrue(schema.id > 0);
+
+      Schema retrieved = getSchema(schema.id, null);
+      assertEquals(schema.id, retrieved.id);
+      assertEquals(schema.uri, retrieved.uri);
+   }
+
+   @org.junit.jupiter.api.Test
+   void testGetSchemaNotFound() {
+      jsonRequest().get("/api/schema/9999")
+          .then()
+          .statusCode(404);
+   }
+   
+   @org.junit.jupiter.api.Test
+   void testGetSchemaByUri() {
+      String schemaUri = "urn:dummy:schema";
+      Schema schema = createSchema("Dummy schema", schemaUri);
+
+      int schemaId = jsonRequest().get("/api/schema/idByUri/" + schemaUri)
+          .then()
+          .statusCode(200)
+          .extract()
+          .as(Integer.class);
+
+      assertEquals(schema.id, schemaId);
+   }
+
+   @org.junit.jupiter.api.Test
+   void testGetSchemaByUriNotFound() {
+      jsonRequest().get("/api/schema/idByUri/urn:unknown:schema")
+          .then()
+          .statusCode(404);
+   }
+
+   @org.junit.jupiter.api.Test
+   void testCreateSchemaWithInvalidId() {
+      Schema schema = new Schema();
+      schema.owner = TESTER_ROLES[0];
+      schema.name = "InvalidSchema";
+      schema.uri = "urn:invalid-id:schema";
+      schema.id = 9999; // does not exist
+
+      jsonRequest().body(schema).post("/api/schema")
+          .then()
+          .statusCode(400);
+   }
+
+
+   @org.junit.jupiter.api.Test
+   void testListSchemasWithDifferentOrderings() {
+      // create some schemas
+      createSchema("Ghi", "urn:schema:ghi");
+      createSchema("Abc", "urn:schema:abc");
+      createSchema("Def", "urn:schema:def");
+
+      // by default order by name and with ascending direction
+      SchemaService.SchemaQueryResult res = listSchemas(null, null, null, null);
+      assertEquals(3, res.schemas.size());
+      assertEquals(3, res.count);
+      assertEquals("Ghi", res.schemas.get(0).name);
+
+      // limit the list to 2 results
+      res = listSchemas(2, 0, null, null);
+      assertEquals(2, res.schemas.size());
+      // total number of records
+      assertEquals(3, res.count);
+   }
+
+   @org.junit.jupiter.api.Test
+   void testDropTokenFromSchema() {
+      // create some schemas
+      Schema s = createSchema("dummy", "urn:dummy:schema", null, "my-super-token");
+
+      SchemaDAO schema = SchemaDAO.findById(s.id);
+      assertNotNull(schema);
+      assertEquals("my-super-token", schema.token);
+
+      jsonRequest().delete("/api/schema/" + schema.id + "/dropToken")
+          .then()
+          .statusCode(204);
+
+      em.clear();
```

Comment:
I am not sure we are not running in a transaction here, so not sure what behaviour `em.clear()` has is this situation.

I think we need to either execute the `.findById()`  calls in separate transactions, or wrap the entire test in a single transaction, and manage the cached result in the test transaction. 

I think 2 separate transactions would be cleaner, as the transaction boundaries are clear and we are not trying to manage entity state change in the DB in a different transaction

Reply:
mmmh shouldn't tests run in a single transaction already or by default?

I found necessary to add `em.clear()` to clear the Hibernate cache and force it to reload the object, without that the changes wouldn't be reflected in the test. Maybe I am missing something else :thinking:

Reply:
As discussed, I will proceed using dedicated transactions for pre and post checks using `Util.withTx` functionality.